include/nbl/builtin/hlsl/bxdf/cook_torrance_base.hlsl

```diff
@@ -35,52 +39,80 @@ struct SCookTorrance
     }
     template<class Query>
-    scalar_type __DG1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(isocache_type) cache)
+    MT __DG1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(isocache_type) cache)
     {
-        return ndf.template DG1<Query>(dg1_query, cache);
+        MT measure_transform;
+        measure_transform.pdf = ndf.template DG1<Query>(query, cache);
+        measure_transform.transmitted = cache.isTransmission();
+        measure_transform.VdotH = cache.getVdotH();
+        measure_transform.LdotH = cache.getLdotH();
+        measure_transform.VdotHLdotH = cache.getVdotHLdotH();
+        return measure_transform;
     }
     template<class Query>
-    scalar_type __DG1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(anisocache_type) cache)
+    MT __DG1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(anisocache_type) cache)
     {
-        return ndf.template DG1<Query>(dg1_query, cache);
+        MT measure_transform;
+        measure_transform.pdf = ndf.template DG1<Query>(query, cache);
+        measure_transform.transmitted = cache.isTransmission();
+        measure_transform.VdotH = cache.getVdotH();
+        measure_transform.LdotH = cache.getLdotH();
+        measure_transform.VdotHLdotH = cache.getVdotHLdotH();
+        return measure_transform;
     }
 
-    // TODO need to make sure ndf functions have the same args -> concept (query, sample, interaction)
     template<class Query>
-    scalar_type __DG(NBL_CONST_REF_ARG(Query) g2_query, NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    MT __DG(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
     {
-        scalar_type DG = ggx_ndf.template D<isocache_type>(cache);
-        if (any<vector<bool, 2> >(A > (vector2_type)numeric_limits<scalar_type>::min))
+        MT measure_transform;
+        measure_transform.pdf = ndf.template D<isocache_type>(cache);
+        NBL_IF_CONSTEXPR(MT::Type == ndf::MicrofacetTransformTypes::MTT_REFLECT_REFRACT)
+            measure_transform.transmitted = cache.isTransmission();
+        NBL_IF_CONSTEXPR(MT::Type == ndf::MicrofacetTransformTypes::MTT_REFRACT)
         {
-            DG *= ggx_ndf.template correlated_wo_numerator<Query, sample_type, isotropic_interaction_type>(g2_query, _sample, interaction);
+            measure_transform.VdotH = cache.getVdotH();
+            measure_transform.LdotH = cache.getLdotH();
+            measure_transform.VdotHLdotH = cache.getVdotHLdotH();
         }
-        return DG;
+        if (any<vector<bool, 2> >(ndf.A > hlsl::promote<vector2_type>(numeric_limits<scalar_type>::min)))
+        {
+            measure_transform.pdf *= ndf.template correlated<Query, sample_type, isotropic_interaction_type>(query, _sample, interaction);
+        }
```

Comment:
explain the reasoning here ?

Reply:
you're trying to see if PDF is not smooth and seeing if G2 has to be applied ?

I'd simply check the `pdf` being `<INF` already and write a comment what its for

Reply:
continued in https://github.com/Devsh-Graphics-Programming/Nabla/pull/916/files#r2310099574

Reply:
continued in https://github.com/Devsh-Graphics-Programming/Nabla/pull/930/files?diff=split&w=1#r2356299166